File: lib/xhpke.dart
// crypto-fl: cryptography primitives and wrappers
// Copyright 2026 Dark Bio AG. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

/// HPKE cryptography wrappers and parametrization.
///
/// https://datatracker.ietf.org/doc/html/rfc9180
///
/// Messages are encrypted to a public key with X-Wing, a hybrid of ML-KEM-768
/// and X25519, and sealed with ChaCha20-Poly1305. Encryption and decryption use
/// an application domain, prefixed with `dark-bio-v1:`, which both sides must
/// agree on. The ciphertext also authenticates a second message that must be
/// supplied separately.
///
/// A [Sender] and [Receiver] pair shares one encapsulated key across many
/// messages, which must be opened in the order they were sealed. The domain is
/// fixed when the pair is created.
///
/// ```dart
/// import 'dart:convert';
///
/// import 'package:darkbio_crypto/xhpke.dart' as xhpke;
///
/// void example() {
///   final secret = xhpke.SecretKey.generate();
///   final domain = utf8.encode('example');
///   final header = utf8.encode('header');
///
///   final (encapKey, ciphertext) = secret.publicKey().seal(msgToSeal: utf8.encode('secret'), msgToAuth: header, domain: domain);
///   final plaintext = secret.open(sessionKey: encapKey, msgToOpen: ciphertext, msgToAuth: header, domain: domain);
///   assert(utf8.decode(plaintext) == 'secret');
///
///   // Many messages under one encapsulated key, opened in the order sealed
///   final (sender, senderKey) = secret.publicKey().newSender(domain: domain);
///   final receiver = secret.newReceiver(encapKey: senderKey, domain: domain);
///   final message = sender.seal(msgToSeal: utf8.encode('first'), msgToAuth: header);
///   assert(utf8.decode(receiver.open(msgToOpen: message, msgToAuth: header)) == 'first');
/// }
/// ```
library;

import 'dart:typed_data';

import 'package:flutter/foundation.dart' show listEquals;

import 'src/generated/api/xhpke.dart' as ffi;

/// An X-Wing private key for decrypting HPKE messages.
class SecretKey {
  final ffi.XhpkeSecretKey _inner;
  SecretKey._(this._inner);

  /// Creates a new, random private key.
  static SecretKey generate() => SecretKey._(ffi.XhpkeSecretKey.generate());

  /// Converts a 32-byte seed into a private key.
  ///
  /// Throws if [bytes] is not 32 bytes long.
  static SecretKey fromBytes(Uint8List bytes) =>
      SecretKey._(ffi.XhpkeSecretKey.fromBytes(bytes: bytes));

  /// Parses a DER buffer into a private key.
  ///
  /// Throws if [der] is not a PKCS #8 v1 encoded X-Wing private key, or if
  /// bytes follow it.
  static SecretKey fromDer(Uint8List der) =>
      SecretKey._(ffi.XhpkeSecretKey.fromDer(der: der));

  /// Parses a PEM string into a private key.
  ///
  /// Throws if [pem] is not a single `PRIVATE KEY` block holding an X-Wing
  /// private key. The block starts at the first byte, uses strict base64 and
  /// ends all its lines in LF or all in CRLF.
  static SecretKey fromPem(String pem) =>
      SecretKey._(ffi.XhpkeSecretKey.fromPem(pem: pem));

  /// Retrieves the public counterpart of the secret key.
  PublicKey publicKey() => PublicKey._(_inner.publicKey());

  /// Returns a 256-bit unique identifier for this key. For HPKE, that is the
  /// SHA256 hash of the raw public key.
  Fingerprint fingerprint() => Fingerprint._(_inner.fingerprint());

  /// Consumes a standalone cryptographic construct encrypted to this secret
  /// key. The method will deconstruct the given encapsulated key and ciphertext
  /// and will also verify the authenticity of the (unencrypted) message-to-auth
  /// (not included in the ciphertext).
  ///
  /// X-Wing is used in HPKE's base mode, which does not authenticate the
  /// sender, so the ciphertext alone does not prove who sent it.
  ///
  /// - [sessionKey]: The 1120-byte encapsulated session key from [PublicKey.seal]
  /// - [msgToOpen]: The ciphertext to decrypt
  /// - [msgToAuth]: Additional authenticated data (must match what was used in seal)
  /// - [domain]: Application domain, the same as the sender's
  ///
  /// Throws if [sessionKey] is not 1120 bytes long, or if decryption fails.
  /// A wrong key, a tampered ciphertext and a mismatched [msgToAuth] or
  /// [domain] all fail alike.
  Uint8List open({
    required Uint8List sessionKey,
    required Uint8List msgToOpen,
    required Uint8List msgToAuth,
    required Uint8List domain,
  }) => _inner.open(
    sessionKey: sessionKey,
    msgToOpen: msgToOpen,
    msgToAuth: msgToAuth,
    domain: domain,
  );

  /// Creates an HPKE receiver context for multi-message decryption using
  /// the encapsulated key from [PublicKey.newSender].
  ///
  /// Messages must be decrypted in the same order they were encrypted. X-Wing
  /// is used in HPKE's base mode, which does not authenticate the sender, so
  /// the context alone does not prove who sent the messages.
  ///
  /// - [encapKey]: The 1120-byte encapsulated key from [PublicKey.newSender]
  /// - [domain]: Application domain, the same as the sender's
  ///
  /// Throws if [encapKey] is not 1120 bytes long or not a valid encapsulated
  /// key.
  Receiver newReceiver({
    required Uint8List encapKey,
    required Uint8List domain,
  }) => Receiver._(_inner.newReceiver(encapKey: encapKey, domain: domain));

  /// Converts a private key into a 32-byte seed.
  Uint8List toBytes() => _inner.toBytes();

  /// Serializes a private key into a DER buffer.
  Uint8List toDer() => _inner.toDer();

  /// Serializes a private key into a PEM string.
  String toPem() => _inner.toPem();
}

/// An X-Wing public key for encrypting HPKE messages.
class PublicKey {
  final ffi.XhpkePublicKey _inner;
  PublicKey._(this._inner);

  /// Converts a 1216-byte array into a public key.
  ///
  /// Throws if [bytes] is not 1216 bytes long, or if a coefficient of its
  /// ML-KEM-768 key falls outside the valid range [0, 3329).
  static PublicKey fromBytes(Uint8List bytes) =>
      PublicKey._(ffi.XhpkePublicKey.fromBytes(bytes: bytes));

  /// Parses a DER buffer into a public key.
  ///
  /// Throws if [der] is not a SubjectPublicKeyInfo encoded X-Wing public key,
  /// or if bytes follow it.
  static PublicKey fromDer(Uint8List der) =>
      PublicKey._(ffi.XhpkePublicKey.fromDer(der: der));

  /// Parses a PEM string into a public key.
  ///
  /// Throws if [pem] is not a single `PUBLIC KEY` block holding an X-Wing
  /// public key. The block starts at the first byte, uses strict base64 and
  /// ends all its lines in LF or all in CRLF.
  static PublicKey fromPem(String pem) =>
      PublicKey._(ffi.XhpkePublicKey.fromPem(pem: pem));

  /// Returns a 256-bit unique identifier for this key. For HPKE, that is the
  /// SHA256 hash of the raw public key.
  Fingerprint fingerprint() => Fingerprint._(_inner.fingerprint());

  /// Creates an HPKE sender context for multi-message encryption.
  ///
  /// Returns the sender context and a 1120-byte encapsulated key that must
  /// be transmitted to the recipient for [SecretKey.newReceiver].
  ///
  /// Messages encrypted with the returned context must be decrypted in order
  /// by the corresponding receiver. X-Wing is used in HPKE's base mode, which
  /// does not authenticate the sender, so the recipient cannot tell who sent
  /// them from the context alone.
  ///
  /// - [domain]: Application domain, the same as the recipient's
  (Sender, Uint8List) newSender({required Uint8List domain}) {
    final (sender, encapKey) = _inner.newSender(domain: domain);
    return (Sender._(sender), encapKey);
  }

  /// Creates a standalone cryptographic construct encrypted to this public
  /// key. The construct will contain the given message-to-seal (encrypted) and
  /// also an authenticity proof for the (unencrypted) message-to-auth (message
  /// not included).
  ///
  /// Returns the 1120-byte encapsulated session key and the ciphertext.
  /// Opening them with [SecretKey.open] needs both, along with [msgToAuth] and
  /// [domain]. X-Wing is used in HPKE's base mode, which does not authenticate
  /// the sender, so the recipient cannot tell who sent the ciphertext from it
  /// alone.
  ///
  /// - [msgToSeal]: The plaintext to encrypt
  /// - [msgToAuth]: Additional authenticated data (not encrypted, but bound)
  /// - [domain]: Application domain, the same as the recipient's
  (Uint8List, Uint8List) seal({
    required Uint8List msgToSeal,
    required Uint8List msgToAuth,
    required Uint8List domain,
  }) => _inner.seal(msgToSeal: msgToSeal, msgToAuth: msgToAuth, domain: domain);

  /// Converts a public key into a 1216-byte array.
  Uint8List toBytes() => _inner.toBytes();

  /// Serializes a public key into a DER buffer.
  Uint8List toDer() => _inner.toDer();

  /// Serializes a public key into a PEM string.
  String toPem() => _inner.toPem();
}

/// A 256-bit unique identifier for an xHPKE key.
class Fingerprint {
  final ffi.XhpkeFingerprint _inner;
  Fingerprint._(this._inner);

  /// Creates a fingerprint from a 32-byte array.
  ///
  /// Throws if [bytes] is not 32 bytes long.
  static Fingerprint fromBytes(Uint8List bytes) =>
      Fingerprint._(ffi.XhpkeFingerprint.fromBytes(bytes: bytes));

  /// Converts a fingerprint into a 32-byte array.
  Uint8List toBytes() => _inner.toBytes();

  /// Whether [other] is an xHPKE fingerprint with the same bytes.
  @override
  bool operator ==(Object other) =>
      other is Fingerprint && listEquals(toBytes(), other.toBytes());

  @override
  int get hashCode => Object.hashAll(toBytes());
}

/// A stateful HPKE sender context for multi-message encryption.
///
/// Each call to [seal] encrypts a message using an auto-incrementing nonce,
/// ensuring unique ciphertexts even for identical plaintexts.
///
/// The corresponding [Receiver] must process messages in the same order
/// they were sealed.
class Sender {
  final ffi.XhpkeSender _inner;
  Sender._(this._inner);

  /// Encrypts a message using the next nonce in the sequence.
  ///
  /// - [msgToSeal]: The plaintext to encrypt
  /// - [msgToAuth]: Additional authenticated data (not encrypted, but bound)
  Uint8List seal({
    required Uint8List msgToSeal,
    required Uint8List msgToAuth,
  }) => _inner.seal(msgToSeal: msgToSeal, msgToAuth: msgToAuth);
}

/// A stateful HPKE receiver context for multi-message decryption.
///
/// Each call to [open] decrypts a message using an auto-incrementing nonce.
/// Messages must be provided in the same order they were sealed by the
/// corresponding [Sender].
class Receiver {
  final ffi.XhpkeReceiver _inner;
  Receiver._(this._inner);

  /// Decrypts a message using the next nonce in the sequence.
  ///
  /// - [msgToOpen]: The ciphertext to decrypt
  /// - [msgToAuth]: Additional authenticated data (must match what was used in seal)
  ///
  /// Throws if decryption fails. A tampered or out-of-order ciphertext and a
  /// mismatched [msgToAuth] all fail alike.
  Uint8List open({
    required Uint8List msgToOpen,
    required Uint8List msgToAuth,
  }) => _inner.open(msgToOpen: msgToOpen, msgToAuth: msgToAuth);
}

/// Exposes the native key to this package's own libraries, not meant for
/// applications.
///
/// @nodoc
extension SecretKeyInternal on SecretKey {
  /// The native key behind this secret key.
  ffi.XhpkeSecretKey get inner => _inner;
}

/// Exposes the native key to this package's own libraries, not meant for
/// applications.
///
/// @nodoc
extension PublicKeyInternal on PublicKey {
  /// The native key behind this public key.
  ffi.XhpkePublicKey get inner => _inner;
}

/// Wraps native fingerprints for this package's own libraries, not meant for
/// applications.
///
/// @nodoc
extension FingerprintInternal on Fingerprint {
  /// Wraps a native fingerprint into a [Fingerprint].
  static Fingerprint wrap(ffi.XhpkeFingerprint inner) => Fingerprint._(inner);
}

File: rust/src/api/xhpke.rs
// crypto-fl: cryptography primitives and wrappers
// Copyright 2026 Dark Bio AG. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

use flutter_rust_bridge::frb;

use super::xdsa::XdsaPublicKey;

/// XhpkeSecretKey is an X-Wing (X25519 + ML-KEM-768) private key for
/// post-quantum hybrid public-key encryption.
#[frb(opaque)]
pub struct XhpkeSecretKey {
    pub(crate) inner: darkbio_crypto::xhpke::SecretKey,
}

impl XhpkeSecretKey {
    /// Generates a new random private key.
    #[frb(sync)]
    pub fn generate() -> Self {
        Self {
            inner: darkbio_crypto::xhpke::SecretKey::generate(),
        }
    }

    /// Creates a private key from a 32-byte seed.
    #[frb(sync)]
    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, String> {
        let bytes_array: [u8; 32] = bytes
            .try_into()
            .map_err(|_| "Invalid key length, expected 32 bytes".to_string())?;
        Ok(Self {
            inner: darkbio_crypto::xhpke::SecretKey::from_bytes(&bytes_array),
        })
    }

    /// Parses a DER-encoded private key.
    #[frb(sync)]
    pub fn from_der(der: Vec<u8>) -> Result<Self, String> {
        Ok(Self {
            inner: darkbio_crypto::xhpke::SecretKey::from_der(&der).map_err(|e| e.to_string())?,
        })
    }

    /// Parses a PEM-encoded private key.
    #[frb(sync)]
    pub fn from_pem(pem: String) -> Result<Self, String> {
        Ok(Self {
            inner: darkbio_crypto::xhpke::SecretKey::from_pem(&pem).map_err(|e| e.to_string())?,
        })
    }

    /// Serializes the private key to a 32-byte seed.
    #[frb(sync)]
    pub fn to_bytes(&self) -> Vec<u8> {
        self.inner.to_bytes().to_vec()
    }

    /// Serializes the private key to DER format.
    #[frb(sync)]
    pub fn to_der(&self) -> Vec<u8> {
        self.inner.to_der()
    }

    /// Serializes the private key to PEM format.
    #[frb(sync)]
    pub fn to_pem(&self) -> String {
        self.inner.to_pem()
    }

    /// Returns the public key corresponding to this private key.
    #[frb(sync)]
    pub fn public_key(&self) -> XhpkePublicKey {
        XhpkePublicKey {
            inner: self.inner.public_key(),
        }
    }

    /// Returns a 32-byte fingerprint uniquely identifying this key.
    #[frb(sync)]
    pub fn fingerprint(&self) -> XhpkeFingerprint {
        XhpkeFingerprint {
            inner: self.inner.fingerprint(),
        }
    }

    /// Creates an HPKE receiver context for multi-message decryption using
    /// the given encapsulated key. Messages must be decrypted in the same
    /// order they were encrypted by the corresponding sender.
    #[frb(sync)]
    pub fn new_receiver(
        &self,
        encap_key: Vec<u8>,
        domain: Vec<u8>,
    ) -> Result<XhpkeReceiver, String> {
        let encap_key_array: [u8; 1120] = encap_key
            .try_into()
            .map_err(|_| "Invalid encapsulated key length, expected 1120 bytes".to_string())?;
        let receiver = self
            .inner
            .new_receiver(&encap_key_array, &domain)
            .map_err(|e| e.to_string())?;
        Ok(XhpkeReceiver { inner: receiver })
    }

    /// Decrypts a message that was encrypted to this key's public counterpart.
    ///
    /// - `session_key`: The 1120-byte encapsulated session key
    /// - `msg_to_open`: The ciphertext to decrypt
    /// - `msg_to_auth`: Additional authenticated data (not encrypted)
    /// - `domain`: Application-specific domain separator
    #[frb(sync)]
    pub fn open(
        &self,
        session_key: Vec<u8>,
        msg_to_open: Vec<u8>,
        msg_to_auth: Vec<u8>,
        domain: Vec<u8>,
    ) -> Result<Vec<u8>, String> {
        let session_key_array: [u8; 1120] = session_key
            .try_into()
            .map_err(|_| "Invalid session key length, expected 1120 bytes".to_string())?;
        self.inner
            .open(&session_key_array, &msg_to_open, &msg_to_auth, &domain)
            .map_err(|e| e.to_string())
    }
}

/// XhpkePublicKey is an X-Wing (X25519 + ML-KEM-768) public key for
/// post-quantum hybrid public-key encryption.
#[frb(opaque)]
pub struct XhpkePublicKey {
    pub(crate) inner: darkbio_crypto::xhpke::PublicKey,
}

impl XhpkePublicKey {
    /// Creates a public key from a 1216-byte array.
    #[frb(sync)]
    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, String> {
        let bytes_array: [u8; 1216] = bytes
            .try_into()
            .map_err(|_| "Invalid key length, expected 1216 bytes".to_string())?;
        Ok(Self {
            inner: darkbio_crypto::xhpke::PublicKey::from_bytes(&bytes_array)
                .map_err(|e| e.to_string())?,
        })
    }

    /// Parses a DER-encoded public key.
    #[frb(sync)]
    pub fn from_der(der: Vec<u8>) -> Result<Self, String> {
        Ok(Self {
            inner: darkbio_crypto::xhpke::PublicKey::from_der(&der).map_err(|e| e.to_string())?,
        })
    }

    /// Parses a PEM-encoded public key.
    #[frb(sync)]
    pub fn from_pem(pem: String) -> Result<Self, String> {
        Ok(Self {
            inner: darkbio_crypto::xhpke::PublicKey::from_pem(&pem).map_err(|e| e.to_string())?,
        })
    }

    /// Parses a public key from a DER-encoded certificate, verifying the xDSA signature.
    /// Returns the key along with validity start and end timestamps (Unix seconds).
    #[frb(sync)]
    pub fn from_cert_der(der: Vec<u8>, signer: &XdsaPublicKey) -> Result<(Self, u64, u64), String> {
        let verified = darkbio_crypto::xhpke::verify_cert_der(
            &der,
            &signer.inner,
            darkbio_crypto::x509::ValidityCheck::Disabled,
        )
        .map_err(|e| e.to_string())?;
        Ok((
            Self {
                inner: verified.public_key,
            },
            verified.cert.not_before,
            verified.cert.not_after,
        ))
    }

    /// Parses a public key from a PEM-encoded certificate, verifying the xDSA signature.
    /// Returns the key along with validity start and end timestamps (Unix seconds).
    #[frb(sync)]
    pub fn from_cert_pem(pem: String, signer: &XdsaPublicKey) -> Result<(Self, u64, u64), String> {
        let verified = darkbio_crypto::xhpke::verify_cert_pem(
            &pem,
            &signer.inner,
            darkbio_crypto::x509::ValidityCheck::Disabled,
        )
        .map_err(|e| e.to_string())?;
        Ok((
            Self {
                inner: verified.public_key,
            },
            verified.cert.not_before,
            verified.cert.not_after,
        ))
    }

    /// Serializes the public key to a 1216-byte array.
    #[frb(sync)]
    pub fn to_bytes(&self) -> Vec<u8> {
        self.inner.to_bytes().to_vec()
    }

    /// Serializes the public key to DER format.
    #[frb(sync)]
    pub fn to_der(&self) -> Vec<u8> {
        self.inner.to_der()
    }

    /// Serializes the public key to PEM format.
    #[frb(sync)]
    pub fn to_pem(&self) -> String {
        self.inner.to_pem()
    }

    /// Generates a DER-encoded X.509 certificate for this public key,
    /// signed by the given xDSA secret key with the specified validity period.
    ///
    /// - `signer`: The xDSA secret key to sign the certificate
    /// - `subject_name`: The subject's common name (CN)
    /// - `issuer_name`: The issuer's common name (CN)
    /// - `not_before`: Certificate validity start time (Unix timestamp)
    /// - `not_after`: Certificate validity end time (Unix timestamp)
    /// - `is_ca`: Whether this is a CA certificate
    /// - `path_len`: Maximum intermediate CAs allowed (only if is_ca is true)
    #[frb(sync)]
    #[allow(clippy::too_many_arguments)]
    pub fn to_cert_der(
        &self,
        signer: &super::xdsa::XdsaSecretKey,
        subject_name: String,
        issuer_name: String,
        not_before: u64,
        not_after: u64,
        is_ca: bool,
        path_len: Option<u8>,
    ) -> Result<Vec<u8>, String> {
        let template = darkbio_crypto::x509::Certificate {
            subject: darkbio_crypto::x509::Name::new().cn(subject_name),
            issuer: darkbio_crypto::x509::Name::new().cn(issuer_name),
            not_before,
            not_after,
            role: if is_ca {
                darkbio_crypto::x509::Role::Authority { path_len }
            } else {
                darkbio_crypto::x509::Role::Leaf
            },
            extensions: Vec::new(),
        };
        darkbio_crypto::xhpke::issue_cert_der(&self.inner, &signer.inner, &template)
            .map_err(|e| e.to_string())
    }

    /// Generates a PEM-encoded X.509 certificate for this public key,
    /// signed by the given xDSA secret key with the specified validity period.
    ///
    /// - `signer`: The xDSA secret key to sign the certificate
    /// - `subject_name`: The subject's common name (CN)
    /// - `issuer_name`: The issuer's common name (CN)
    /// - `not_before`: Certificate validity start time (Unix timestamp)
    /// - `not_after`: Certificate validity end time (Unix timestamp)
    /// - `is_ca`: Whether this is a CA certificate
    /// - `path_len`: Maximum intermediate CAs allowed (only if is_ca is true)
    #[frb(sync)]
    #[allow(clippy::too_many_arguments)]
    pub fn to_cert_pem(
        &self,
        signer: &super::xdsa::XdsaSecretKey,
        subject_name: String,
        issuer_name: String,
        not_before: u64,
        not_after: u64,
        is_ca: bool,
        path_len: Option<u8>,
    ) -> Result<String, String> {
        let template = darkbio_crypto::x509::Certificate {
            subject: darkbio_crypto::x509::Name::new().cn(subject_name),
            issuer: darkbio_crypto::x509::Name::new().cn(issuer_name),
            not_before,
            not_after,
            role: if is_ca {
                darkbio_crypto::x509::Role::Authority { path_len }
            } else {
                darkbio_crypto::x509::Role::Leaf
            },
            extensions: Vec::new(),
        };
        darkbio_crypto::xhpke::issue_cert_pem(&self.inner, &signer.inner, &template)
            .map_err(|e| e.to_string())
    }

    /// Returns a 32-byte fingerprint uniquely identifying this key.
    #[frb(sync)]
    pub fn fingerprint(&self) -> XhpkeFingerprint {
        XhpkeFingerprint {
            inner: self.inner.fingerprint(),
        }
    }

    /// Creates an HPKE sender context for multi-message encryption to this
    /// public key. Returns the sender context and the 1120-byte encapsulated
    /// key that must be transmitted to the recipient.
    ///
    /// Messages encrypted with the returned context must be decrypted in order
    /// by the corresponding receiver context.
    #[frb(sync)]
    pub fn new_sender(&self, domain: Vec<u8>) -> Result<(XhpkeSender, Vec<u8>), String> {
        let (sender, encap_key) = self.inner.new_sender(&domain).map_err(|e| e.to_string())?;
        Ok((XhpkeSender { inner: sender }, encap_key.to_vec()))
    }

    /// Encrypts a message to this public key.
    ///
    /// Returns a tuple of (1120-byte session key, ciphertext).
    ///
    /// - `msg_to_seal`: The plaintext to encrypt
    /// - `msg_to_auth`: Additional authenticated data (not encrypted)
    /// - `domain`: Application-specific domain separator
    #[frb(sync)]
    pub fn seal(
        &self,
        msg_to_seal: Vec<u8>,
        msg_to_auth: Vec<u8>,
        domain: Vec<u8>,
    ) -> Result<(Vec<u8>, Vec<u8>), String> {
        let (session_key, ciphertext) = self
            .inner
            .seal(&msg_to_seal, &msg_to_auth, &domain)
            .map_err(|e| e.to_string())?;
        Ok((session_key.to_vec(), ciphertext))
    }
}

/// XhpkeSender wraps an HPKE sender encryption context for multi-message
/// communication. Each call to `seal` encrypts a message using an
/// auto-incrementing nonce, ensuring unique ciphertexts even for identical
/// plaintexts.
///
/// The corresponding `XhpkeReceiver` must process messages in the same order
/// they were sealed.
#[frb(opaque)]
pub struct XhpkeSender {
    inner: darkbio_crypto::xhpke::Sender,
}

impl XhpkeSender {
    /// Encrypts a message using the next nonce in the sequence.
    #[frb(sync)]
    pub fn seal(&mut self, msg_to_seal: Vec<u8>, msg_to_auth: Vec<u8>) -> Result<Vec<u8>, String> {
        self.inner
            .seal(&msg_to_seal, &msg_to_auth)
            .map_err(|e| e.to_string())
    }
}

/// XhpkeReceiver wraps an HPKE receiver decryption context for multi-message
/// communication. Each call to `open` decrypts a message using an
/// auto-incrementing nonce.
///
/// Messages must be provided in the same order they were sealed by the
/// corresponding `XhpkeSender`.
#[frb(opaque)]
pub struct XhpkeReceiver {
    inner: darkbio_crypto::xhpke::Receiver,
}

impl XhpkeReceiver {
    /// Decrypts a message using the next nonce in the sequence.
    #[frb(sync)]
    pub fn open(&mut self, msg_to_open: Vec<u8>, msg_to_auth: Vec<u8>) -> Result<Vec<u8>, String> {
        self.inner
            .open(&msg_to_open, &msg_to_auth)
            .map_err(|e| e.to_string())
    }
}

/// XhpkeFingerprint is a 32-byte unique identifier for an XHPKE key.
#[frb(opaque)]
pub struct XhpkeFingerprint {
    pub(crate) inner: darkbio_crypto::xhpke::Fingerprint,
}

impl XhpkeFingerprint {
    /// Creates a fingerprint from a 32-byte array.
    #[frb(sync)]
    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, String> {
        let bytes_array: [u8; 32] = bytes
            .try_into()
            .map_err(|_| "Invalid fingerprint length, expected 32 bytes".to_string())?;
        Ok(Self {
            inner: darkbio_crypto::xhpke::Fingerprint::from_bytes(&bytes_array),
        })
    }

    /// Serializes the fingerprint to a 32-byte array.
    #[frb(sync)]
    pub fn to_bytes(&self) -> Vec<u8> {
        self.inner.to_bytes().to_vec()
    }
}
